internal/rapid/reflect/type.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reflect
+
+import (
+	"fmt"
+	"reflect"
+
+	"pgregory.net/rapid"
+)
+
+type GenerateType = *rapid.Generator[reflect.Type]
+
+var pulumiFieldName = rapid.StringMatching("[a-zA-Z]+")
+var structFieldName = rapid.StringMatching("[A-Z][a-zA-Z]*")
+
+func String() GenerateType { return rapid.Just(reflect.TypeOf("")) }
+func Bool() GenerateType   { return rapid.Just(reflect.TypeOf(false)) }
+func Number() GenerateType { return rapid.Just(reflect.TypeOf(float64(0.0))) }
+func Null() GenerateType   { return rapid.Just(reflect.TypeOf(nil)) }
+func Struct(maxDepth int) GenerateType {
+	return rapid.Custom(func(t *rapid.T) reflect.Type {
+		if maxDepth <= 0 {
+			t.Fatalf("Cannot create struct with maxDepth <= 0")
+		}
+		numFields := rapid.IntRange(1, 8).Draw(t, "numFields")
+		fieldNames := rapid.SliceOfNDistinct(
+			structFieldName, numFields, numFields, rapid.ID[string],
+		).Draw(t, "fieldNames")
+		pulumiNames := rapid.SliceOfNDistinct(
+			pulumiFieldName, numFields, numFields, rapid.ID[string],
+		).Draw(t, "pulumiNames")
+		structFields := make([]reflect.StructField, numFields)
+		for i := range fieldNames {
+			var optional string
+			typ := Type(maxDepth - 1)
+
+			// It's not worth it to round trip values through ptr types. We
+			// make sure that all types are optional.
+			//
+			// We might want to relax this later, since we do allow non-ptr
+			// optional types.
+			if rapid.Bool().Draw(t, fmt.Sprintf("optional-%d", i)) {
```

Comment:
I don't quite follow this comment. I can see that we're only generating pointers as optional, but what does it mean that "It's not worth it to round trip values through ptr types"?

Reply:
I have changed the comment so it's easier to read:
https://github.com/pulumi/pulumi-go-provider/blob/df0520f704274237ec670d8e0c5a3da8a8568167/internal/rapid/reflect/type.go#L50-L72